### PR TITLE
[DO NOT MERGE] Smart type mapper

### PR DESCRIPTION
**Pull request by Kagamine**
Making type mapper auto detect the connstr and mysql version.

Related issues: #137, #145

### CODE REVIEW

File: src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlTypeMapper.cs [hunk 1]
```diff
@@ -60,9 +68,27 @@ public class MySqlTypeMapper : RelationalTypeMapper
...
+            this.services = services;
+            var relationalConn = services.GetService<MySqlRelationalConnection>();
+            var TreatTinyIntAsBoolean = true;
+            var UseDateTime6 = false;
+            if (relationalConn != null)
+            {
+                var csb = new MySqlConnectionStringBuilder(relationalConn.DbConnection.ConnectionString);
+                TreatTinyIntAsBoolean = csb.TreatTinyAsBoolean;
+                relationalConn.DbConnection.Open();
+                using (var cmd = new MySqlCommand("SELECT VERSION();", relationalConn.DbConnection as MySqlConnection))
```
Comment (caleblloyd) [inline src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlTypeMapper.cs]: relationalConn.DbConnection.ServerVersion should be populated after a connection has been opened [api reference](https://docs.microsoft.com/en-us/dotnet/core/api/system.data.common.dbconnection#System_Data_Common_DbConnection_ServerVersion)

We don't want to require a connection to be opened here in non-scaffolding cases though.  Any way to make the version check depend on scaffolding?

Comment (Kagamine) [inline src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlTypeMapper.cs]: datetime(6) is related to migrations part.

Comment (caleblloyd) [inline src/Pomelo.EntityFrameworkCore.MySql/Storage/Internal/MySqlTypeMapper.cs]: Oh, if TypeMapper is only loaded in Migrations and Scaffolding then it should be OK to open a connection here.  As long as it does not get loaded during normal operation.

**Comment by Kagamine** [comment]
image

It seems that we cannot get the service while scaffolding.

**Comment by caleblloyd** [comment]
> It seems that we cannot get the service while scaffolding.

I know it may be bad design, but we could create a `static` class that saves `MySqlRelationalConnection` metadata the first time a connection is opened in a process and use that.

This would not take into account multiple connection strings in the same process.  But I don't think that ever happens with scaffolding and migrations.

**Comment by Kagamine** [comment]
No, it will be a bad design when using ef to connect multi databases

**Pull request closed without merging**